tests/email/test_dm_mailchimp.py

```diff
@@ -10,298 +11,301 @@
-        create_or_update.side_effect = RequestException("error sending", response=response)
+            assert res is False
+            assert log_catcher.records[1].msg == 'Mailchimp failed to add user (foo) to list (list_id)'
+            assert log_catcher.records[1].error == "error sending"
+            assert log_catcher.records[1].levelname == 'ERROR'
 
-        res = dm_mailchimp_client.subscribe_new_email_to_list('list_id', 'example@example.com')
+    def test_subscribe_new_emails_to_list(self):
+        dm_mailchimp_client = DMMailChimpClient('username', 'api key', mock.MagicMock())
+        with mock.patch.object(dm_mailchimp_client, 'subscribe_new_email_to_list', autospec=True):
+            dm_mailchimp_client.subscribe_new_email_to_list.return_value = True
 
-        assert res is True
-        logger.error.assert_called_once_with(
-            "Expected error: Mailchimp failed to add user (foo) to list (list_id). API error: The email address looks fake or invalid, please enter a real email address.",  # noqa
-            extra={"error": "error sending"}
-        )
+            with assert_external_service_log_entry(count=2):
+                res = dm_mailchimp_client.subscribe_new_emails_to_list('list_id',
+                                                                       ['email1@example.com', 'email2@example.com'])
+            calls = [mock.call('list_id', 'email1@example.com'), mock.call('list_id', 'email2@example.com')]
 
+            assert res is True
+            dm_mailchimp_client.subscribe_new_email_to_list.assert_has_calls(calls)
 
-@mock.patch("logging.Logger", autospec=True)
-@mock.patch("dmutils.email.dm_mailchimp.DMMailChimpClient.get_email_hash", return_value="foo")
-def test_handles_responses_with_invalid_json(get_email_hash, logger):
-    dm_mailchimp_client = DMMailChimpClient('username', 'api key', logger)
-    with mock.patch.object(
-            dm_mailchimp_client._client.lists.members, 'create_or_update', autospec=True) as create_or_update:
-        response = mock.Mock()
-        response.json.side_effect = JSONDecodeError('msg', 'doc', 0)
-        create_or_update.side_effect = RequestException("error sending", response=response)
+    def test_subscribe_new_emails_to_list_tries_all_emails_returns_false_on_error(self):
+        dm_mailchimp_client = DMMailChimpClient('username', 'api key', mock.MagicMock())
+        with mock.patch.object(
+                dm_mailchimp_client, 'subscribe_new_email_to_list', autospec=True) as subscribe_new_email_to_list:
+            subscribe_new_email_to_list.side_effect = [False, True]
 
-        res = dm_mailchimp_client.subscribe_new_email_to_list('list_id', 'example@example.com')
-
-        assert res is False
-        assert logger.error.call_args == mock.call(
-            'Mailchimp failed to add user (foo) to list (list_id)', extra={'error': 'error sending'}
-        )
+            with assert_external_service_log_entry(count=2):
```

Comment:
Do we need to assert anything from log_catcher here? There are a few other examples below (for both expected success and expected failure) where we capture logs with the context manager but don't check the log message content. Or are we not too bothered about that?

Reply:
The `assert_external_service_log_entry` actually performs assertions automatically for the external service log messages. If you need to make _extra_ assertions (e.g. for other/additional log messages), you can store the log catcher in a variable to get access to all of the log records that are caught. I've done this in a few of the tests to allow them to make their old/existing assertions for other log messages.